Update Hovercards style

jcheringer opened the pull request:
## Proposed Changes

This PR switches the hovercards styles from the current wide format to a portrait design.
This new format will give users more space to display important profile elements, such as calls-to-action (CTAs), URLs, verified badges, and the new Hero image.

https://github.com/user-attachments/assets/bd4cbbdb-1647-40ac-a218-b9b228fe3d92

## Testing Instructions

* Checkout this branch and, using your terminal, navigate to `/web/packages/hovercards`
* Run `npm run start` - the playground window should open
* Play with the hovercards in the playground and check if they work as expected
* If necessary, you can change the hovercards data here:
  *  `web/packages/hovercards/playground/core.ts`
  * `web/packages/hovercards/playground/index.html`

jcheringer committed:
Update Hovercards style

jcheringer committed:
Improve code structure

jcheringer committed:
General adjustments

jcheringer committed:
Update hovercards package version

jcheringer committed:
Add nonce support

Sopheap-cloud commented on web/packages/hovercards/playground/core.ts:
```diff
@@ -6,12 +6,13 @@ import { Hovercards } from '../dist';
-		// To test the empty about me case
-		myHash: '99c3338797c95c418d9996bd39931506',
-		i18n: {
-			'View profile': 'View profile 😜',
```
Ok

jcheringer committed:
Add inline styles through JS

wellyshen commented on web/packages/hovercards/package.json:
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "@gravatar-com/hovercards",
-	"version": "0.9.3",
+	"version": "0.10.0",
```
The GH release action will update the version for us.

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -20,6 +20,30 @@ export interface VerifiedAccount {
 	isHidden: boolean;
 }
 
+export interface ContactInfo {
```
Not Mandatory: For a type with all optional properties, I usually use `Partial<>` to reduce thinking effort for the props in future development 😆, similar to how we handled the [Options](https://github.com/Automattic/gravatar/blob/trunk/web/packages/hovercards/src/core.ts#L78) type.

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -295,32 +366,227 @@ export default class Hovercards {
+
+		const hovercardInner = hovercard.querySelector( '.gravatar-hovercard__inner' );
+		const headerImageEl = hovercardInner.querySelector< HTMLElement >( '.gravatar-hovercard__header-image' );
+		const profileColorEl = hovercardInner.querySelector< HTMLElement >( '.gravatar-hovercard__profile-color' );
```
Nit: Consider specifying a more precise element type:

```suggestion
		const headerImageEl = hovercardInner.querySelector< HTMLDivElement >( '.gravatar-hovercard__header-image' );
		const profileColorEl = hovercardInner.querySelector< HTMLDivElement >( '.gravatar-hovercard__profile-color' );
```

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -295,32 +366,227 @@ export default class Hovercards {
+	 * @param {Element} container - The container context to search for the drawer.
+	 * @return {void}
+	 */
+	private static _openDrawer( selector: string, container: Element ): void {
```
Nit: Consider specifying a more precise element type:

```suggestion
	private static _openDrawer( selector: string, container: HTMLElement ): void {
```

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -295,32 +366,227 @@ export default class Hovercards {
+				<div class="gravatar-hovercard__drawer-backdrop backdrop"></div>
+				<div class="gravatar-hovercard__drawer-card">
+					<div class="gravatar-hovercard__drawer-header">
+						<h1 class="gravatar-hovercard__drawer-title">${ titleText }</h1>
```
I recommend using `<h2>` or lower-level headings to avoid unnecessary conflicts with screen readers on a page. This approach aligns with best practices followed by popular UI component libraries, such as [Material-UI](https://mui.com/material-ui/react-dialog/) and the [W3Schools example](https://www.w3schools.com/w3css/w3css_modal.asp).

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -295,32 +366,227 @@ export default class Hovercards {
+	 * @param {string} id        - The drawer id
+	 * @param {string} titleText - The title shown at the drawer's header
+	 * @param {string} content   - The drawer inner content
+	 * @return {string}          - The drawer HTML string
```
Nit:

```suggestion
	 * @param {string} id        - The drawer id.
	 * @param {string} titleText - The title shown at the drawer's header.
	 * @param {string} content   - The drawer inner content.
	 * @return {string}          - The drawer HTML string.
```

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -295,32 +366,227 @@ export default class Hovercards {
+	 */
+	private static _createDrawer( id: string, titleText: string, content: string ): string {
+		return `
+			<div id="${ id }" class="gravatar-hovercard__drawer">
```
Q: What is the `id` used for?

jcheringer replied:
It is used to target the necessary drawer more easily when handling events.

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -295,32 +366,227 @@ export default class Hovercards {
 						<span class="gravatar-hovercard__profile-link-text">
-							${ isEditProfile ? __( i18n, 'Edit your profile' ) : __( i18n, 'View profile' ) }
+							${ isEditProfile ? __( i18n, 'Edit your profile →' ) : __( i18n, 'View profile →' ) }
 						</span>
```
Since we have used the `→` to replace the SVG icon. We can remove the `<span>` tag.

wellyshen commented on web/packages/hovercards/src/style.scss:
```diff
@@ -97,30 +102,74 @@ $color-blue: #1d4fc4;
 
 	.gravatar-hovercard__profile-link {
 		display: flex;
 		align-items: center;
```
We can remove the `align-items: center`  since we have replaced the SVG icon with the `→`.

wellyshen commented on web/packages/hovercards/src/style.scss:
```diff
@@ -97,30 +102,74 @@ $color-blue: #1d4fc4;
 		display: flex;
 		align-items: center;
 		text-decoration: none;
 		gap: 4px;
```
Also the `gap: 4px`.

wellyshen commented on web/packages/hovercards/src/style.scss:
```diff
@@ -4,6 +4,8 @@ $color-super-light-gray: #eee;
 $color-light-gray: #d8dbdd;
 $color-gray: #707070;
 $color-blue: #1d4fc4;
+/* stylelint-disable value-keyword-case */
```
Disable the rule for the next line only:

```suggestion
/* stylelint-disable-next-line value-keyword-case */
```

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -394,6 +660,12 @@ export default class Hovercards {
 			}
 
+			const urlParams = new URLSearchParams( params );
+			if ( ! urlParams.has( 's' ) ) {
+				urlParams.append( 's', '128' );
+				params = `?${ urlParams.toString() }`;
+			}
```
Thank you for the perf improvement 💙. We also need to consider the following cases:
- Consider using double image size for Retina displays (e.g. `256`)
- The URL might be:
  - `https://0.gravatar.com/avatar/123?size=128`
  - `https://0.gravatar.com/avatar/123?d=identicon`
  - And more... (please refer to [this doc](https://docs.gravatar.com/api/avatars/images/))

Additionally, if I recall correctly, the [avatarUrl](https://github.com/Automattic/gravatar/blob/trunk/web/packages/hovercards/src/core.ts#L403) is only used for the hovercard. Therefore, we could also set the existing `s` or `size` parameter to `128`.

jcheringer replied:
I'm not sure I understood the suggested change here. Are you saying we should set the `s` or `size` to 128 regardless of the presence of this parameter on the `params` variable?

wellyshen replied:
@jcheringer Sorry, for making the confusion 😅.

Yes, the [avatarUrl](https://github.com/Automattic/gravatar/blob/trunk/web/packages/hovercards/src/core.ts#L403) is used exclusively by the hovercard. Therefore, we can set it to 256 (regardless of the existing s or size parameters) to provide better resolution and support for Retina displays:

| `s=128` (less clear) | `s=256` (more clear) |
| - | - |
| <img width="102" alt="截圖 2024-11-21 下午1 48 57" src="https://github.com/user-attachments/assets/f64f6230-ee3a-4a51-9151-a843085b9213"> | <img width="118" alt="截圖 2024-11-21 下午1 50 29" src="https://github.com/user-attachments/assets/971aba56-8177-4979-9d81-506ec35fb9b6"> |

My idea:

```ts
const urlParams = new URLSearchParams( params );
urlParams.delete( 's' );
urlParams.delete( 'size' );
urlParams.append( 's', '256' );
params = `?${ urlParams.toString() }`;
```

wellyshen replied:
@jcheringer Sorry, for making the confusion 😅.

Yes, the [avatarUrl](https://github.com/Automattic/gravatar/blob/trunk/web/packages/hovercards/src/core.ts#L403) is used exclusively by the hovercard. Therefore, we can set it to `256` (regardless of the existing `s` or `size` parameters) to provide better resolution and support for Retina displays:

| `s=128` (less clear) | `s=256` (more clear) |
| - | - |
| <img width="102" alt="截圖 2024-11-21 下午1 48 57" src="https://github.com/user-attachments/assets/f64f6230-ee3a-4a51-9151-a843085b9213"> | <img width="118" alt="截圖 2024-11-21 下午1 50 29" src="https://github.com/user-attachments/assets/971aba56-8177-4979-9d81-506ec35fb9b6"> |

My idea:

```ts
const urlParams = new URLSearchParams( params );
urlParams.delete( 'size' );
urlParams.set( 's', '256' );
params = `?${ urlParams.toString() }`;
```

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -295,32 +366,227 @@ export default class Hovercards {
+			Hovercards._openDrawer( `#${ contactsDrawerId }`, hovercardInner );
+		} );
+
+		hovercardInner.querySelector( `#${ contactsDrawerId } .close-btn` )?.addEventListener( 'click', () => {
+			Hovercards._closeDrawer( `#${ contactsDrawerId }`, hovercardInner );
+		} );
+
+		hovercardInner.querySelector( `#${ contactsDrawerId } .backdrop` )?.addEventListener( 'click', () => {
+			Hovercards._closeDrawer( `#${ contactsDrawerId }`, hovercardInner );
+		} );
```
I'm not sure if I've missed any context, but I think using `.gravatar-hovercard__drawer-backdrop` and `.gravatar-hovercard__drawer-close` should be sufficient.

jcheringer replied:
I'm not sure I understood the suggestion.
Could you clarify your suggestion with a code snippet?
If I change it to:

```JavaScript
		hovercardInner.querySelector( '.gravatar-hovercard__drawer-close' )?.addEventListener( 'click', () => {
			Hovercards._closeDrawer( `#${ contactsDrawerId }`, hovercardInner );
		} );
```

I would target only the first available drawer.

wellyshen replied:
Sorry, I mean:

```diff
- hovercardInner.querySelector( `#${ contactsDrawerId } .close-btn` )
+ hovercardInner.querySelector( `#${ contactsDrawerId } .gravatar-hovercard__drawer-close` )
```

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -295,32 +366,227 @@ export default class Hovercards {
+
+			return `
+				<li class="gravatar-hovercard__drawer-item">
+					<img class="gravatar-hovercard__drawer-item-icon" width="24" height="24" src="${ `https://secure.gravatar.com/${ icons[ key ] }` }" alt="">
```
Should we give the empty `alt` a value?

jcheringer replied:
What is your suggestion in this case?
From my perspective, the icon is a [decorative image](https://www.w3.org/WAI/tutorials/images/decorative/), that's why I used the empty alt text.

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -295,32 +366,227 @@ export default class Hovercards {
+		payments.links?.forEach( ( item ) => {
+			items.push( `
+				<li class="gravatar-hovercard__drawer-item">
+					<img class="gravatar-hovercard__drawer-item-icon" width="24" height="24" src="https://secure.gravatar.com/icons/link.svg" alt="">
```
Should we give the empty `alt` a value?

jcheringer replied:
Same as https://github.com/Automattic/gravatar/pull/123#discussion_r1850486084

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -295,32 +366,227 @@ export default class Hovercards {
+	 * Creates the contact drawer content.
+	 *
+	 * @param {Record< string, any >[]} contactsData - The user's contact data
+	 * @return {string}                              - The contact drawer content
```
Nit:

```suggestion
	 * @param {Record< string, any >[]} contactsData - The user's contact data.
	 * @return {string}                              - The contact drawer content.
```

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -295,32 +366,227 @@ export default class Hovercards {
+			let text = value.replace( /^(https?:\/\/)/, '' );
+
+			if ( url ) {
+				text = `<a href="${ url }" target="_blank">${ text }</a>`;
```
Could we add a class to this `<a>` (e.g. `gravatar-hovercard__drawer-item-link`) so developers can override its style more easily?

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -295,32 +366,227 @@ export default class Hovercards {
+	 * Creates the send money drawer content.
+	 *
+	 * @param {Payments} payments - The user's payment data
+	 * @return {string}           - The send money drawer content
```
Nit: 
```suggestion
	 * @param {Payments} payments - The user's payment data.
	 * @return {string}           - The send money drawer content.
```

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -295,32 +366,227 @@ export default class Hovercards {
+					return `mailto:${ value }`;
+				case 'contact_form':
+				case 'calendar':
+					return value.startsWith( 'http' ) ? value : null;
```
Q: Could there be other cases like the following?

<img width="411" alt="截圖 2024-11-20 下午5 46 52" src="https://github.com/user-attachments/assets/288e9ee9-7e96-4f73-86ea-3775414e8d47">

jcheringer replied:
Good catch. I have changed the logic to display the links correctly.

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -295,32 +366,227 @@ export default class Hovercards {
+					<div class="gravatar-hovercard__drawer-item-info">
+						<span class="gravatar-hovercard__drawer-item-label">${ item.label }</span>
+						<span class="gravatar-hovercard__drawer-item-text">
+							<a href="${ item.url }" target="_blank">${ item.url.replace( /^(https?:\/\/)/, '' ) }</a>
```
Could we add a class to this `<a>` (e.g. gravatar-hovercard__drawer-item-link) so developers can override its style more easily?

wellyshen commented on web/packages/hovercards/src/index.ts:
```diff
@@ -1,6 +1,8 @@
 export type { Placement } from './compute-position';
 export type {
 	VerifiedAccount,
+	ContactInfo,
+	Payments,
```
We usually export all the possible types to enable developers to handle any cases they might needed:

```suggestion
	ContactInfo,
	Payments,
	PaymentLink,
	CryptoWallet
```

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -506,6 +779,9 @@ export default class Hovercards {
 			hovercard.style[ padding ] = `${ paddingValue }px`;
 
+			// Placing the hovercard at the top-level of the dc to avoid being clipped by overflow
+			dc.body.appendChild( hovercard );
```
This change will cause the hovercard to be positioned incorrectly. By default, the [placement](https://github.com/Automattic/gravatar/tree/trunk/web/packages/hovercards#placement-string--right) of a hovercard is set to [right](https://github.com/Automattic/gravatar/blob/trunk/web/packages/hovercards/src/core.ts#L130).

Additionally, the hovercard's position is determined by its `width` and `height`. Therefore, it is crucial to ensure that the sizes of all cards (i.e. the skeleton card, error card, and hovercard) are consistent. Otherwise, the hovercard's position may be incorrect on the first hover.

However, ensuring the skeleton card and the hovercard are the same `height` might be challenging since the skeleton card is displayed before the data is fetched. We might need to consider an alternative (maybe remove it?) 🤔

<img width="430" alt="截圖 2024-11-20 下午6 00 54" src="https://github.com/user-attachments/assets/f66e2189-bb48-41b0-bf0b-76135119c035">

jcheringer replied:
I have reverted this change and updated the hovercard skeleton to match the new hovercard structure better.
However, because some elements are shown according to the user's preferences, it will be tricky to make the hovercard height 100% accurate.
From my perspective, with the latest changes, the layout shift after the load is acceptable.

wellyshen commented on web/packages/hovercards/src/style.scss:
```diff
@@ -46,20 +39,32 @@ $color-blue: #1d4fc4;
-		align-items: center;
-		gap: 10px;
+		flex-direction: column;
+		align-items: flex-start;
```
Q: Do we need `align-items: flex-start`?

wellyshen commented on web/packages/hovercards/src/style.scss:
```diff
@@ -46,20 +39,32 @@ $color-blue: #1d4fc4;
 	}
 
 	.gravatar-hovercard__header {
+		position: relative;
```
Nit: I might have missed some context. It seems that `z-index: 1` is more suitable for this case.

wellyshen commented on web/packages/hovercards/src/style.scss:
```diff
@@ -187,3 +244,129 @@ $color-blue: #1d4fc4;
+		text-transform: capitalize;
+	}
+
+	.gravatar-hovercard__drawer-item-text a {
```
Could we assign a class (e.g., `.gravatar-hovercard__drawer-item-link`) to the element and use it as the selector?

wellyshen commented on web/packages/hovercards/src/style.scss:
```diff
@@ -187,3 +244,129 @@ $color-blue: #1d4fc4;
+	.gravatar-hovercard__drawer-item {
+		display: flex;
+		align-items: flex-start;
+		justify-content: flex-start;
```
Q: Do we need the `align-items: flex-start` and `justify-content: flex-start`?

wellyshen reviewed (commented):
Thank you for contributing to the OSS. The architecture looks good to me! I have left some comments, please take a look 👀

- Regarding the skeleton card and the error card, I feel they don’t align well with the hovercard. This is particularly noticeable when hovering over an avatar, as the layout shift can be quite jarring. From a UX perspective, I think it would be ideal for the `height` of these cards to match, ensuring a smoother transition. (BTW, I’m not a designer. Should we consult @Luchadores?)
  - Also related to [this issue](https://github.com/Automattic/gravatar/pull/123#discussion_r1850011646)

https://github.com/user-attachments/assets/8d597e31-3cea-4da4-a1cb-eaa967138c00

(Apart from the size, we could also adjust the number and position slightly)
<img width="316" alt="截圖 2024-11-20 下午3 43 00" src="https://github.com/user-attachments/assets/376453a4-0e6b-4af5-bed8-7ffadb86a83e">

- Note: We need to review and update the relevant sections (e.g., Image, API, Translation, etc.) of the [README.md](https://github.com/Automattic/gravatar/blob/trunk/web/packages/hovercards/README.md) before releasing this package.

jcheringer committed:
Address feedback

jcheringer committed:
Address feedback

jcheringer committed:
Adjust hovercard skeleton

jcheringer committed:
Playground update

jcheringer committed:
Update docs

jcheringer committed:
Update README.md

jcheringer commented:
Thanks for the review, @wellyshen!
I have addressed most of the feedback and added some answers/questions.
Could you take a look again?

> Regarding the skeleton card and the error card, I feel they don't align well with the hovercard. This is particularly noticeable when hovering over an avatar, as the layout shift can be quite jarring. From a UX perspective, I think it would be ideal for the height of these cards to match, ensuring a smoother transition. 

I added a comment about this [here](https://github.com/Automattic/gravatar/pull/123#discussion_r1850862568).

> Note: We need to review and update the relevant sections (e.g., Image, API, Translation, etc.) of the [README.md](https://github.com/Automattic/gravatar/blob/trunk/web/packages/hovercards/README.md) before releasing this package.

I also updated the docs. Let me know if I missed anything.

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -278,13 +314,48 @@ export default class Hovercards {
 
 		hovercard.innerHTML = `
-			<div class="gravatar-hovercard__inner">
+			<div id="g-${ hash }" class="gravatar-hovercard__inner">
```
Q: What is the `id="g-${ hash }"` used for?

jcheringer replied:
That was a leftover from when I tried to address the CSP issue. Good catch!

wellyshen commented on web/packages/hovercards/README.md:
```diff
@@ -489,8 +493,8 @@ A few items of text are used when displaying a hovercard. The library contains E
 
-- `Edit your profile`
-- `View profile`
+- `Edit your profile →`
+- `View profile →`
```
Two more:

```suggestion
- `Edit your profile →`
- `View profile →`
- `Contact`
- `Send money`
```

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -295,32 +366,229 @@ export default class Hovercards {
+	 * Opens a hovercard drawer.
+	 *
+	 * @param {string}  selector  - The drawer selector.
+	 * @param {Element} container - The container context to search for the drawer.
```
Nit:

```suggestion
	 * @param {HTMLElement} container - The container context to search for the drawer.
```

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -295,32 +366,229 @@ export default class Hovercards {
+	 * @param {Element} container - The container context to search for the drawer.
+	 * @return {void}
+	 */
+	private static _closeDrawer( selector: string, container: Element ): void {
```
Nit: Consider using a more specific element type:

```suggestion
	 * @param {string}  selector  - The drawer selector.
	 * @param {HTMLElement} container - The container context to search for the drawer.
	 * @return {void}
	 */
	private static _closeDrawer( selector: string, container: HTMLElement ): void {
```

wellyshen commented on web/packages/hovercards/src/core.ts:
```diff
@@ -295,32 +366,229 @@ export default class Hovercards {
+		const items = contactsData.map( ( [ key, value ]: string[] ) => {
+			const url = getUrl( key, value );
+			let text = value.replace( /^(https?:\/\/)/, '' );
+			text = text.replace( /^(www\.)/, '' );
```
It could be a single `replace` to reduce the bundle size, if you want:

```suggestion
			const text = value.replace( /^(https?:\/\/)?(www\.)?/, '' );
```

wellyshen reviewed (commented):
> Note:  It seems we need to support auto-shift for the new portrait hovercard. I will address this in [this issue](https://github.com/Automattic/gravatar/issues/135).

Thank you for addressing the feedback! I have taken another look and left some comments. Please take a look 👀.

### 🔴 The position of the hovercard appears incorrect due to inconsistencies in the `height` between the skeleton card and the hovercard. Here, I use the auto-flip case as an example:
#### To address this issue, we could:
- Ensure that the `height` of all cards is consistent.
- Alternatively, remove the skeleton card.
- Another option would be to update the positioning logic to account for this scenario (this might be the most complex solution).
- Any other ideas?

https://github.com/user-attachments/assets/1b83ccf3-a547-4df8-8452-acf546f4f166

### 🟡 In the case of multiple profile cards, there are duplicate element IDs in the document, which could potentially cause unexpected issues in the future:

<img width="1800" alt="截圖 2024-11-21 下午3 02 17" src="https://github.com/user-attachments/assets/d5e07346-970f-4d10-b724-8a4316ffac7e">

#### Here's a possible approach to avoid using element IDs while achieving a smaller bundle size and more concise code:

Before:
<img width="731" alt="截圖 2024-11-21 下午4 00 35" src="https://github.com/user-attachments/assets/95b72042-69e3-47b4-b968-ce808a4858aa">
After:
<img width="750" alt="截圖 2024-11-21 下午4 00 15" src="https://github.com/user-attachments/assets/0f266b4e-d02e-4cd1-924f-21befa788bac">

```ts
static createHovercard: CreateHovercard = ( profileData, { additionalClass, myHash, i18n = {} } = {} ) => {
	// ...

	if ( nonEmptyContacts.length ) {
		ctaButtons += `
			<button class="gravatar-hovercard__button" data-target-drawer="contact">${ __( i18n, 'Contact' ) }</button>
		`;

		contactsDrawer = Hovercards._createDrawer(
			'contact',
			__( i18n, 'Contacts' ),
			Hovercards._createContactDrawerContent( nonEmptyContacts )
		);
	}

	if ( hasPayments ) {
		ctaButtons += `
			<button class="gravatar-hovercard__button" data-target-drawer="send-money">${ __( i18n, 'Send money' ) }</button>
		`;

		sendMoneyDrawer = Hovercards._createDrawer(
			'send-money',
			__( i18n, 'Send money' ),
			Hovercards._createSendMoneyDrawerContent( payments )
		);
	}

	// ...

	hovercardInner.querySelectorAll( '.gravatar-hovercard__button' ).forEach( ( el: HTMLButtonElement ) => {
		el.addEventListener( 'click', () => Hovercards._openDrawer( el, hovercardInner ) );
	} );
	hovercardInner.querySelectorAll( '.gravatar-hovercard__drawer-close' ).forEach( ( el: HTMLButtonElement ) => {
		el.addEventListener( 'click', () => Hovercards._closeDrawer( el, hovercardInner ) );
	} );
	hovercardInner.querySelectorAll( '.gravatar-hovercard__drawer-backdrop' ).forEach( ( el: HTMLDivElement ) => {
		el.addEventListener( 'click', () => Hovercards._closeDrawer( el, hovercardInner ) );
	} );

	return hovercard;
};

private static _createDrawer( name: string, titleText: string, content: string ): string {
	return `
		<div class="gravatar-hovercard__drawer" data-drawer-name="${ name }">
			<div class="gravatar-hovercard__drawer-backdrop" data-target-drawer="${ name }"></div>
			// ...
			<button class="gravatar-hovercard__drawer-close" data-target-drawer="${ name }">
				// ...
			</button>
			// ...	
		</div>
	`;
}

private static _openDrawer( target: HTMLElement, container: HTMLElement ): void {
	const drawer = container.querySelector( `.gravatar-hovercard__drawer[data-drawer-name="${ target.dataset.targetDrawer }"]` );

	if ( ! drawer ) {
		return;
	}

	drawer.classList.add( 'gravatar-hovercard__drawer--open' );
}

private static _closeDrawer( target: HTMLElement, container: HTMLElement ): void {
	const drawer = container.querySelector( `.gravatar-hovercard__drawer[data-drawer-name="${ target.dataset.targetDrawer }"]` );

	if ( ! drawer ) {
		return;
	}

	drawer.classList.add( 'gravatar-hovercard__drawer--closing' );
	drawer.classList.remove( 'gravatar-hovercard__drawer--open' );

	setTimeout( () => {
		drawer.classList.remove( 'gravatar-hovercard__drawer--closing' );
	}, 300 );
}
```

jcheringer commented:
> *  Ensure that the height of all cards is consistent.

Because users can know show/hide some elements on the hovercard, we don't have a way to know the height before loading the user data.

> * Alternatively, remove the skeleton card.

What does that mean? 

> * Another option would be to update the positioning logic to account for this scenario (this might be the most complex solution).

I don't see other options other than trying to work with the positioning logic.

jcheringer committed:
Address feedback

jcheringer committed:
Address feedback

jcheringer committed:
Recalculate position after fetching data

jcheringer committed:
Clean up

jcheringer commented:
@wellyshen, recalculating the card position after fetching the user data is a possible solution to ensure the card position is correct.
To achieve that, I slightly changed the function `computingPosition` to avoid code duplication.
Please take a look and let me know your thoughts.

jcheringer committed:
Adjust verified icon

wellyshen commented on web/packages/hovercards/src/compute-position.ts:
The hovercard's position is slightly off during the first hover:

| 1st hovering | 2nd hovering |
| - | - |
| <img width="346" alt="截圖 2024-11-22 下午4 34 03" src="https://github.com/user-attachments/assets/b1c07f04-0134-4e1c-9511-888601925d8a"> | <img width="346" alt="截圖 2024-11-22 下午4 33 53" src="https://github.com/user-attachments/assets/2587c37f-55e7-4211-ac49-7f113afb8944"> |

Here’s the updated version of the function. Additionally, since we have made modifications to the function, updating its name and documentation for better maintainability would be great!

- New file name: `assign-position.ts`

```diff 
 /**
- * Computes the position of a card relative to a ref element.
+ * Compute and assign the position of the card element.
  *
  * @param {HTMLElement}    ref          - The ref element.
  * @param {HTMLDivElement} card         - The card element.
  * @param {Options}        [options={}] - The placement, offset, and auto-flip options.
  * @return {void}
  */
-export default function computingPosition(
+export default function assignPosition(
 	ref: HTMLElement,
 	card: HTMLDivElement,
 	{ placement = 'right', offset = 0, autoFlip = true }: Options = {}
 ): void {
+	// Reset the card's padding for re-calculation.
+	card.style.padding = '0';
+
 	const refRect = ref.getBoundingClientRect();
 	const cardRect = card.getBoundingClientRect();
 	const refScrollT = refRect.top + scrollY;

```

jcheringer replied:
Is this really necessary?
From my perspective, the function is still the same:  computing the card position.

jcheringer replied:
I have changed it anyway.

wellyshen commented on web/packages/hovercards/src/style.scss:
```diff
@@ -97,49 +102,81 @@ $color-blue: #1d4fc4;
+		min-height: 42px;
 		display: flex;
 		align-items: center;
-		text-decoration: none;
-		gap: 4px;
+		justify-content: center;
```
Q: Do we need the `display`, `align-items`, and `justify-content` for `<button>` elements? It works fine without these props.

wellyshen commented on web/packages/hovercards/src/style.scss:
```diff
@@ -179,11 +222,135 @@ $color-blue: #1d4fc4;
+		border-top-right-radius: 4px;
+		padding: 20px 0;
+		box-sizing: border-box;
+		transform: translateY(100%);
```
Utilize the GPU to ensure smooth animations:

```suggestion
		transform: translate3d(0, 100%, 0);
```

wellyshen commented on web/packages/hovercards/src/style.scss:
```diff
@@ -179,11 +222,135 @@ $color-blue: #1d4fc4;
+		}
+
+		.gravatar-hovercard__drawer-card {
+			transform: translateY(0);
```
Utilize the GPU to ensure smooth animations:

```suggestion
			transform: translate3d(0, 0, 0);
```

wellyshen commented on web/packages/hovercards/src/style.scss:
```diff
@@ -179,11 +222,135 @@ $color-blue: #1d4fc4;
+		}
+
+		.gravatar-hovercard__drawer-card {
+			transform: translateY(100%);
```
Utilize the GPU to ensure smooth animations:

```suggestion
			transform: translate3d(0, 100%, 0);
```

wellyshen reviewed (commented):
Again, thank you for addressing the feedback. I think we're getting close to merging the PR. I have left some comments, please take a look 👀.

- Should we set the [default placement](https://github.com/Automattic/gravatar/tree/trunk/web/packages/hovercards#placement-string--right) to `right-start` to better align with the UX of the new portrait hovercards?

jcheringer committed:
Make sure to call onHovercardShow once the hovercard is loaded

jcheringer committed:
Address feedback

wellyshen reviewed (approved):
LGTM, great job bro 💯. Let's ship it 🚀
In addition, I have implemented [the auto-shift feature](https://github.com/Automattic/gravatar/pull/136), please take a look, thank you!

Note: We also need to update [this developer doc](https://docs.gravatar.com/sdk/web-components/).
